src/core/store-data.ts

```diff
@@ -1,56 +1,223 @@
+    owner,
+    repo,
+    sha: branch,
+    per_page: 1,
+  });
+
+  return {
+    commitSha: latestCommit.sha,
+    treeSha: latestCommit.commit.tree.sha,
+  };
+}
 
 export const storeData = async (metricData: MetricData[]) => {
-  const errored: string[] = [];
+  if (metricData.length === 0) {
+    // this is just a bug in our code and doesn't get a message in LogErrors
+    throw new Error("storeData must be called with at least one dataset");
+  }
+
+  const owner = getRequiredEnvVar("REPO_OWNER");
+  const repo = getRequiredEnvVar("REPO_NAME");
+  const repoPath = getRequiredEnvVar("REPO_PATH");
+  const targetBranch = getRequiredEnvVar("TARGET_BRANCH");
+  const retriesAllowed = getOptionalNumericEnvVar("CONFLICT_RETRIES") ?? 0;
+
+  // Build the tree of added files from the metric data
+  const fileTree = metricData.map((data) => ({
+    path: `${repoPath}/${data.owner}/${data.repo}/${data.metric_signature}/${data.created_at}.json`,
+    mode: "100644" as const,
+    // UTF-8 encoded plaintext files can be sent as a plain string.
+    content: `${JSON.stringify(data)}\n`,
+  }));
+
+  let headShas: { commitSha: string; treeSha: string };
 
-  for (const data of metricData) {
-    logger.pending(
-      `Pushing file to repo: ${process.env.REPO_PATH}/${data.created_at}.json`
+  /*
+    [Step 1/4]
+    
+    Retrieve current head commit SHA of branch.
+
+    Since this is the first read from the repo, we'll throw a custom error
+    warning about the token not having read access in case the request fails.
+  */
+  try {
+    headShas = await getHeadShas(owner, repo, targetBranch);
+  } catch (e: unknown) {
+    if (e instanceof Error && "status" in e && typeof e.status === "number") {
+      throw StoreDataError.noReadAccess(e.status);
+    } else {
+      throw e;
+    }
+  }
+
+  // Note: the actual exit condition for a retry is a bit more complex than just
+  // this check, see the last request inside the for loop where the code either
+  // throws an error, breaks the loop, or advances it.
+  for (let tryNumber = 0; tryNumber <= retriesAllowed; tryNumber++) {
+    /*
+      [Step 2/4]
+      
+      Post the git tree object to GitHub, so we can reference it inside a new
+      commit. It uses the current HEAD's tree as the base tree.
+
+      Since this is the first write to the repo, we'll throw a custom error
+      warning about the token not having write access in case the request fails.
+    */
+    const {
+      data: { sha: newTreeSha },
+    } = await octokitForDataRepo
+      .request("POST /repos/{owner}/{repo}/git/trees", {
+        owner,
+        repo,
+        base_tree: headShas.treeSha,
+        tree: fileTree,
+      })
+      .catch((e: unknown) => {
+        if (
+          e instanceof Error &&
+          "status" in e &&
+          typeof e.status === "number"
+        ) {
+          throw StoreDataError.noWriteAccess(e.status);
+        } else {
+          throw e;
+        }
+      });
+
+    // ASSUMPTION:
```

Comment:
for now it is fine to make that assumption as they are all from the same github event